### PR TITLE
Fix DBSCAN handling of non-core points

### PR DESCRIPTION
**Pull request by rcurtin**
This handles #3339.  @iad-ABDUL-RAOUF, thanks for reporting the issue!  If you are willing to review the changes here and see if they make sense (at least the comments for the approach), I would appreciate it.  I think I have done it correctly but I may have dropped a small detail.

The problem is that the existing DBSCAN implementation grows clusters "through" noise/non-core points (defined as points that have less than the minimum number of neighbors `minPoints`).  This is demonstrated by the nice test case that @iad-ABDUL-RAOUF supplied.  The fix essentially boils down to allowing clusters to add non-core points, but *not* connect two disparate clusters through a noise point.

Our DBSCAN implementation strategy differs a good deal from the original algorithm's pseudocode and uses a union find structure to process points serially.  I spent a while considering it, and to the best of my understanding our implementation will give the same result as the original algorithm, although it does look quite different.

### CODE REVIEW

File: src/mlpack/methods/dbscan/dbscan_impl.hpp [hunk 1]
```diff
@@ -154,18 +155,71 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
...
-    for (size_t j = 0; j < neighbors[0].size(); ++j)
-      uf.Union(i, neighbors[0][j]);
+    // Union to all neighbors if the point is not noise.
+    //
+    // Note that the +1 is needed because our range search will return the point
+    // itself as within the range.
+    //
+    // If the point is noise, we leave its label as undefined (i.e. we do no
+    // unioning).
+    if (neighbors[0].size() >= minPoints + 1)
```
Comment (AdarshSantoria) [inline src/mlpack/methods/dbscan/dbscan_impl.hpp:195]: I think +1 will not be there as choosen point will also lie inside the circle of eps acorss that point.
For the datset - 
	1	2	2	8	8	25
	2	2	3	7	8	80
and eps=3 and minpts=2.
This shows label -
        0      0      0      SIZE_MAX      SIZE_MAX      SIZE_MAX      
But actual label is -
        0      0      0      1     1     SIZE_MAX 

Example SRC - First example in https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html

File: src/mlpack/methods/dbscan/dbscan_impl.hpp [hunk 2]
```diff
@@ -100,7 +100,8 @@ size_t DBSCAN<RangeSearchType, PointSelectionPolicy>::Cluster(
     arma::Row<size_t>& assignments)
 {
   // Initialize the UnionFind object.
-  UnionFind uf(data.n_cols);
+  // The last element represents noise.
+  UnionFind uf(data.n_cols + 1);
```
Comment (iad-ABDUL-RAOUF) [inline src/mlpack/methods/dbscan/dbscan_impl.hpp:104]: The last element of this "data.n_cols + 1" sized UnionFind representing noise seems unused.

**Review by iad-ABDUL-RAOUF — CHANGES_REQUESTED** [review]
Thanks @rcurtin for the hard work. The implementation is different from the original DBSCAN, so I had some trouble to convince myself it was really the same (especially the non-batch algorithm). But after thinking about it, I think it looks good.

The main issue I have is with minPoints (see the detailed commentaries I left).

I am also curious about whether this implementation is faster or slower than the original algorithm. But I guess it is difficult to assess for now...

File: src/mlpack/methods/dbscan/dbscan_impl.hpp (hunk 2 above)
Comment (rcurtin) [reply src/mlpack/methods/dbscan/dbscan_impl.hpp:104]: Yes, thank you!  This was a change I made during a draft, and forgot to change it back.  Reverted in c8d7a9575b.

File: src/mlpack/methods/dbscan/dbscan_impl.hpp (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/methods/dbscan/dbscan_impl.hpp]: Thanks, you are correct, and you saw exactly why I misunderstood.  Using the reasoning in your footnote, I removed the `+ 1` throughout in 8e1f30ba58.  Note that this means for the dual-tree search, we use `- 1` now, as dual-tree monochromatic range search will not return the point in its own range (similar to k-nearest-neighbor search).

File: src/mlpack/methods/dbscan/dbscan_impl.hpp [hunk 3]
```diff
@@ -180,21 +234,43 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::BatchCluster(
...
   // Now loop over all points.
   for (size_t i = 0; i < data.n_cols; ++i)
   {
     // Get the next index.
     const size_t index = pointSelector.Select(i, data);
-    for (size_t j = 0; j < neighbors[index].size(); ++j)
-      uf.Union(index, neighbors[index][j]);
+    // Monochromatic dual-tree range search does not return the point as its own
+    // neighbor, so we do not need to add 1 to `minPoints`.
+    if (neighbors[index].size() >= minPoints)
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/dbscan/dbscan_impl.hpp]: Thanks---solved and responded to [here](https://github.com/mlpack/mlpack/pull/3346#discussion_r1062611478). :+1:

**Review by zoq — APPROVED** [review]
Looks good to me.

**Comment by iad-ABDUL-RAOUF** [comment]
Sorry guys, i'm extremely busy with my phd right now. I won't be able to check the updates until march... Don't wait for me...
